pkg/agent/controller/networkpolicy/fqdn_test.go

```diff
@@ -584,3 +591,152 @@ func TestSyncDirtyRules(t *testing.T) {
+
+	tests := []struct {
+		name                  string
+		existingDNSCache      map[string]dnsMeta
+		dnsResponseIPs        map[string]ipWithExpiration
+		expectedIPs           map[string]ipWithExpiration
+		expectedRequeryAfter  *time.Duration
+		mockSelectorToRuleIDs map[fqdnSelectorItem]sets.Set[string]
+	}{
+		{
+			name: "new IP added",
+			existingDNSCache: map[string]dnsMeta{
+				testFQDN: {
+					responseIPs: map[string]ipWithExpiration{
+						"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+						"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(6 * time.Second)},
+					},
+				},
+			},
+			dnsResponseIPs: map[string]ipWithExpiration{
+				"192.1.1.3": {ip: net.ParseIP("192.1.1.3"), expirationTime: currentTime.Add(10 * time.Second)},
+			},
+			expectedIPs: map[string]ipWithExpiration{
+				"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+				"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(6 * time.Second)},
+				"192.1.1.3": {ip: net.ParseIP("192.1.1.3"), expirationTime: currentTime.Add(10 * time.Second)},
+			},
+			expectedRequeryAfter: ptr.To(5 * time.Second),
+		},
+		{
+			name: "empty DNS response",
+			existingDNSCache: map[string]dnsMeta{
+				testFQDN: {
+					responseIPs: map[string]ipWithExpiration{
+						"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+						"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(6 * time.Second)},
+					},
+				},
+			},
+			dnsResponseIPs: map[string]ipWithExpiration{},
+			expectedIPs: map[string]ipWithExpiration{
+				"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+				"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(6 * time.Second)},
+			},
+		},
+		{
+			name: "old IP present in DNS response is retained with an updated TTL fetched from response",
+			existingDNSCache: map[string]dnsMeta{
+				testFQDN: {
+					responseIPs: map[string]ipWithExpiration{
+						"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+						"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(1 * time.Second)},
+					},
+				},
+			},
+			dnsResponseIPs: map[string]ipWithExpiration{
+				"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(1 * time.Second)},
+				"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(5 * time.Second)},
+			},
+			expectedIPs: map[string]ipWithExpiration{
+				"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+				"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(5 * time.Second)},
+			},
+			expectedRequeryAfter: ptr.To(5 * time.Second),
+		},
+		{
+			name: "stale IP with expired TTL is evicted",
```

Comment:
This unit test has failed in CI, please look into fixing it:
```
--- FAIL: TestOnDNSResponse (1.24s)
    --- FAIL: TestOnDNSResponse/stale_IP_with_expired_TTL_is_evicted (1.00s)
        fqdn_test.go:732: 
            	Error Trace:	/home/runner/work/antrea/antrea/pkg/agent/controller/networkpolicy/fqdn_test.go:732
            	Error:      	Condition never satisfied
            	Test:       	TestOnDNSResponse/stale_IP_with_expired_TTL_is_evicted
```

Reply:
This is a tricky failure. I actually had to solve a similar problem a couple of months ago: https://github.com/antrea-io/antrea/pull/6407
You can use the same fix here.
In essence, the problem is that using a fake clock alongside a delaying workqueue is quite error prone. There is the possibility for a race when advancing the clock with `Step`, while the workqueue implementation is processing a newly added item and creates a timer.

Reply:
@antoninbas Thanks for the guidance on this one. Requesting your review. One thing I did differently here was avoid re-implementing the wrapper around ```fakeClock``` and its methods. Since the current test is in the same package as the wrapper, and it's returning a pointer to the```fakeClock``` instance which should not be a problem with multiple tests using it.